File: frontend/src/types/resource-analytics.ts
/**
 * Resource Analytics Type Definitions
 * Comprehensive types for resource utilization, cost optimization, and efficiency metrics
 */

export type TimeFrame = '24h' | '7d' | '30d' | '90d';
export type Granularity = 'hourly' | 'daily' | 'weekly';
export type ResourceType = 'compute' | 'tokens' | 'api' | 'storage' | 'network' | 'cost';
export type WasteType = 'idle_resources' | 'oversized_instances' | 'redundant_api_calls' |
  'inefficient_prompts' | 'unused_cache' | 'failed_execution_costs' |
  'rate_limit_waste' | 'token_overflow';
export type OptimizationCategory = 'compute' | 'tokens' | 'api' | 'storage' | 'network' | 'cost' | 'general';
export type BudgetStatus = 'no_budget' | 'within_budget' | 'warning' | 'critical' | 'over_budget';
export type ImplementationEffort = 'low' | 'medium' | 'high';
export type Priority = 'low' | 'medium' | 'high' | 'critical';

// ============================================================================
// Compute Resources
// ============================================================================
export interface CPUUsage {
  averagePercent: number;
  peakPercent: number;
  coreSeconds: number;
}

export interface MemoryUsage {
  averageMb: number;
  peakMb: number;
  allocationMb: number;
  averageAllocationMb?: number;
}

export interface GPUUsage {
  utilizationPercent: number;
  memoryMb: number;
  computeUnits: number;
  totalComputeUnits?: number;
  averageUtilizationPercent?: number;
  maxMemoryMb?: number;
}

export interface NetworkIO {
  bytesSent: number;
  bytesReceived: number;
  apiCalls: number;
  totalBytesTransferred?: number;
  totalBytesSent?: number;
  totalBytesReceived?: number;
  totalApiCalls?: number;
}

export interface ComputeMetrics {
  cpuUsage: CPUUsage;
  memoryUsage: MemoryUsage;
  gpuUsage?: GPUUsage;
  networkIo: NetworkIO;
  executionMetrics?: {
    totalExecutions: number;
    avgDurationMs: number;
    avgQueueWaitMs: number;
  };
}

// ============================================================================
// Token Metrics
// ============================================================================
export interface TokenDistribution {
  inputTokens: number;
  outputTokens: number;
  embeddingTokens: number;
  inputPercent: number;
  outputPercent: number;
  embeddingPercent: number;
}

export interface TokenAverages {
  tokensPerExecution: number;
  inputTokensPerExecution: number;
  outputTokensPerExecution: number;
}

export interface TokenEfficiency {
  tokensPerDollar: number;
  cacheHitRate: number;
  totalCacheHits: number;
  tokensPerSecond?: number;
}

export interface TokenCost {
  totalCostUsd: number;
  avgCostPerExecution: number;
  costPerThousandTokens: number;
}

export interface ModelUsage {
  provider: string;
  model: string;
  totalTokens: number;
  usageCount: number;
  costUsd: number;
}

export interface TokenMetrics {
  totalTokens: number;
  tokenDistribution: TokenDistribution;
  averages: TokenAverages;
  efficiency: TokenEfficiency;
  cost: TokenCost;
  modelUsage: ModelUsage[];
}

// ============================================================================
// API Metrics
// ============================================================================
export interface APICallStatistics {
  totalCalls: number;
  successfulCalls: number;
  failedCalls: number;
  rateLimitedCalls: number;
  timeoutCalls?: number;
}

export interface APILatencyMetrics {
  avgLatencyMs: number;
  p50LatencyMs: number;
  p95LatencyMs: number;
  p99LatencyMs: number;
  minLatencyMs: number;
  maxLatencyMs: number;
}

export interface APICostMetrics {
  totalCostUsd: number;
  costPerCall: number;
  wastedCostUsd: number;
  wastedCostFailedCalls?: number;
}

export interface APIBreakdown {
  endpoint: string;
  provider: string | null;
  totalCalls: number;
  successfulCalls: number;
  failedCalls: number;
  rateLimitedCalls: number;
  avgLatencyMs: number;
  p95LatencyMs: number;
  costUsd: number;
  errorRate: number;
}

export interface APIMetrics {
  totalApiCalls: number;
  successRate: number;
  statistics: APICallStatistics;
  cost: APICostMetrics;
  apiBreakdown: APIBreakdown[];
}

// ============================================================================
// Storage Metrics
// ============================================================================
export interface StorageMetrics {
  tempStorage: {
    averageMb: number;
    maxMb: number;
  };
  persistentStorage: {
    averageMb: number;
    maxMb: number;
  };
  cache: {
    averageSizeMb: number;
    maxSizeMb: number;
  };
  database: {
    totalOperations: number;
  };
  cost: {
    totalCostUsd: number;
  };
}

// ============================================================================
// Cost Analytics
// ============================================================================
export interface CostByCategory {
  computeCostUsd: number;
  tokenCostUsd: number;
  apiCostUsd: number;
  storageCostUsd: number;
  networkCostUsd: number;
}

export interface CostDistribution {
  computePercent: number;
  tokenPercent: number;
  apiPercent: number;
  storagePercent: number;
  networkPercent: number;
}

export interface CostBreakdown {
  totalCostUsd: number;
  costByCategory: CostByCategory;
  costDistribution: CostDistribution;
  averageCostPerExecution: number;
  totalExecutions: number;
}

// ============================================================================
// Efficiency Metrics
// ============================================================================
export interface EfficiencyMetrics {
  overallScore: number;
  tokensPerDollar: number;
  executionsPerDollar: number;
  throughputScore: number;
  costEfficiencyPercent: number;
  totalWasteCostUsd: number;
  percentileRank?: number;
  improvementAreas?: string[];
}

// ============================================================================
// Resource Usage Response
// ============================================================================
export interface ResourceUsage {
  agentId: string;
  workspaceId: string;
  timeframe: TimeFrame;
  periodStart: string;
  periodEnd: string;
  computeMetrics: ComputeMetrics;
  tokenMetrics: TokenMetrics;
  apiMetrics: APIMetrics;
  storageMetrics: StorageMetrics;
  costBreakdown: CostBreakdown;
  efficiencyMetrics: EfficiencyMetrics;
}

// ============================================================================
// Optimization
// ============================================================================
export interface OptimizationRecommendation {
  id?: string;
  type: string;
  category: OptimizationCategory;
  priority: Priority;
  title: string;
  description: string;
  reasoning?: string;
  currentValue?: number;
  recommendedValue?: number;
  estimatedSavingsUsd: number;
  estimatedSavingsPercent?: number;
  estimatedSavings?: string;
  implementationEffort: ImplementationEffort;
  implementationSteps?: string[];
  status?: string;
  createdAt?: string;
}

export interface CostOptimization {
  currentMonthlyCost: number;
  potentialSavings: number;
  optimizationRecommendations: OptimizationRecommendation[];
}

// ============================================================================
// Waste Detection
// ============================================================================
export interface WasteEvent {
  id: string;
  agentId: string;
  executionId?: string;
  workspaceId: string;
  wasteType: WasteType;
  wasteCategory: string;
  title: string;
  description: string;
  wasteAmount: number;
  wasteUnit: string;
  wasteCostUsd: number;
  confidenceScore: number;
  isResolved: boolean;
  detectedAt: string;
}

export interface WasteSummary {
  totalWasteCostUsd: number;
  wasteByType: Record<string, number>;
  unresolvedCount: number;
  potentialMonthlySavings: number;
}

// ============================================================================
// Forecasting
// ============================================================================
export interface ResourceForecast {
  resourceType: ResourceType;
  forecastHorizonDays: number;
  predictedValue: number;
  lowerBound: number;
  upperBound: number;
  confidenceLevel: number;
  confidence?: number;
  historicalAverage: number;
  historicalTrend: string;
  exceedsBudget: boolean;
  forecastDate: string;
  projectedValue?: number;
}

export interface CostProjection {
  projectedDailyCost?: number;
  projectedWeeklyCost?: number;
  projectedMonthlyCost: number;
  lowerBound?: number;
  upperBound?: number;
  growthRate: number;
  budgetAlerts: string[];
}

export interface ForecastResponse {
  tokenUsage: ResourceForecast | null;
  computeUsage: ResourceForecast | null;
  projectedCosts: CostProjection | null;
  budgetAlerts: string[];
}

// ============================================================================
// Token Analysis
// ============================================================================
export interface TokenAnalysisResponse {
  tokenDistribution: {
    inputTokens: number;
    outputTokens: number;
    embeddingTokens: number;
    totalTokens: number;
    contextUtilization: {
      averageUsed: number;
      maxUsed: number;
    };
  };
  efficiencyMetrics: {
    tokensPerSecond: number;
    cacheHitRate: number;
    tokensPerDollar: number;
    totalCacheHits: number;
  };
  optimizationOpportunities: OptimizationRecommendation[];
  costAnalysis: {
    totalCost: number;
    totalTokens: number;
    avgCostPerDay: number;
    trendData: Array<{
      date: string;
      tokens: number;
      cost: number;
    }>;
  };
}

// ============================================================================
// Comprehensive Analytics
// ============================================================================
export interface ResourceAnalytics {
  agentId: string;
  workspaceId: string;
  timeframe: TimeFrame;
  tokenAnalysis: TokenAnalysisResponse;
  costOptimization: CostOptimization;
  wasteDetection: WasteSummary;
  forecasts: ForecastResponse;
}

// ============================================================================
// Budget Management
// ============================================================================
export interface TokenBudget {
  workspaceId: string;
  agentId?: string;
  dailyTokenBudget?: number;
  weeklyTokenBudget?: number;
  monthlyTokenBudget?: number;
  dailyCostBudgetUsd?: number;
  weeklyCostBudgetUsd?: number;
  monthlyCostBudgetUsd?: number;
  warningThresholdPercent: number;
  criticalThresholdPercent: number;
}

export interface BudgetUsage {
  dailyTokens: number;
  dailyCostUsd: number;
  dailyBudgetUsagePercent: number;
  dailyCostBudgetUsagePercent: number;
  rolling7dTokens: number;
  rolling30dTokens: number;
  rolling30dCostUsd: number;
  budgetStatus: BudgetStatus;
}

// ============================================================================
// Workspace Analytics
// ============================================================================
export interface WorkspaceCostAnalysis {
  workspaceId: string;
  period: string;
  totalCost: number;
  costByAgent: Record<string, number>;
  costBreakdown: CostByCategory & { totalCostUsd: number };
  costTrends?: Array<{
    date: string;
    cost: number;
  }>;
  projections?: CostProjection;
  optimizationPotential?: number;
}

// ============================================================================
// Efficiency Leaderboard
// ============================================================================
export interface EfficiencyLeaderboardEntry {
  agentId: string;
  rank: number;
  efficiencyScore: number;
  tokensPerDollar: number;
  executionsPerDollar: number;
  costEfficiencyPercent: number;
  monthlyCost: number;
}

export interface EfficiencyLeaderboard {
  workspaceId: string;
  metric: string;
  leaderboard: EfficiencyLeaderboardEntry[];
}

// ============================================================================
// Chart Data Types
// ============================================================================
export interface TimeSeriesDataPoint {
  timestamp: string;
  value: number;
  label?: string;
}

export interface CostTrendData {
  date: string;
  compute: number;
  tokens: number;
  api: number;
  storage: number;
  network: number;
  total: number;
}

export interface TokenUsageData {
  date: string;
  inputTokens: number;
  outputTokens: number;
  embeddingTokens: number;
  totalTokens: number;
}

File: backend/src/models/schemas/resource_analytics.py
"""Comprehensive resource utilization and analytics schemas."""

from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import datetime, date
from enum import Enum


# ============================================================================
# Enums
# ============================================================================
class ResourceType(str, Enum):
    """Resource types for tracking."""
    COMPUTE = "compute"
    TOKENS = "tokens"
    API = "api"
    STORAGE = "storage"
    NETWORK = "network"
    COST = "cost"


class WasteType(str, Enum):
    """Types of resource waste."""
    IDLE_RESOURCES = "idle_resources"
    OVERSIZED_INSTANCES = "oversized_instances"
    REDUNDANT_API_CALLS = "redundant_api_calls"
    INEFFICIENT_PROMPTS = "inefficient_prompts"
    UNUSED_CACHE = "unused_cache"
    FAILED_EXECUTION_COSTS = "failed_execution_costs"
    RATE_LIMIT_WASTE = "rate_limit_waste"
    TOKEN_OVERFLOW = "token_overflow"


class OptimizationCategory(str, Enum):
    """Categories for optimization recommendations."""
    COMPUTE = "compute"
    TOKENS = "tokens"
    API = "api"
    STORAGE = "storage"
    NETWORK = "network"
    COST = "cost"
    GENERAL = "general"


class BudgetStatus(str, Enum):
    """Budget status indicators."""
    NO_BUDGET = "no_budget"
    WITHIN_BUDGET = "within_budget"
    WARNING = "warning"
    CRITICAL = "critical"
    OVER_BUDGET = "over_budget"


class ImplementationEffort(str, Enum):
    """Implementation effort levels."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class Priority(str, Enum):
    """Priority levels."""
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


# ============================================================================
# Compute Resource Schemas
# ============================================================================
class CPUUsage(BaseModel):
    """CPU usage metrics."""
    averagePercent: float = Field(..., alias="averagePercent", description="Average CPU utilization percentage")
    peakPercent: float = Field(..., alias="peakPercent", description="Peak CPU utilization percentage")
    coreSeconds: float = Field(..., alias="coreSeconds", description="Total core-seconds consumed")

    class Config:
        populate_by_name = True


class MemoryUsage(BaseModel):
    """Memory usage metrics."""
    averageMb: float = Field(..., alias="averageMb", description="Average memory usage in MB")
    peakMb: float = Field(..., alias="peakMb", description="Peak memory usage in MB")
    allocationMb: float = Field(..., alias="allocationMb", description="Total allocated memory in MB")

    class Config:
        populate_by_name = True


class GPUUsage(BaseModel):
    """GPU usage metrics."""
    utilizationPercent: float = Field(..., alias="utilizationPercent", description="GPU utilization percentage")
    memoryMb: float = Field(..., alias="memoryMb", description="GPU memory used in MB")
    computeUnits: float = Field(..., alias="computeUnits", description="GPU compute units consumed")

    class Config:
        populate_by_name = True


class NetworkIO(BaseModel):
    """Network I/O metrics."""
    bytesSent: int = Field(..., alias="bytesSent", description="Total bytes sent")
    bytesReceived: int = Field(..., alias="bytesReceived", description="Total bytes received")
    apiCalls: int = Field(..., alias="apiCalls", description="Total API calls made")

    class Config:
        populate_by_name = True


class ComputeResources(BaseModel):
    """Compute resource metrics."""
    cpuUsage: CPUUsage = Field(..., alias="cpuUsage")
    memoryUsage: MemoryUsage = Field(..., alias="memoryUsage")
    gpuUsage: Optional[GPUUsage] = Field(None, alias="gpuUsage")
    networkIo: NetworkIO = Field(..., alias="networkIo")

    class Config:
        populate_by_name = True


# ============================================================================
# LLM Resource Schemas
# ============================================================================
class LLMResources(BaseModel):
    """LLM resource usage metrics."""
    model: str = Field(..., description="Model name")
    modelProvider: str = Field(..., alias="modelProvider", description="Model provider")
    inputTokens: int = Field(..., alias="inputTokens", description="Input tokens consumed")
    outputTokens: int = Field(..., alias="outputTokens", description="Output tokens generated")
    embeddingTokens: int = Field(0, alias="embeddingTokens", description="Embedding tokens used")
    totalTokens: int = Field(..., alias="totalTokens", description="Total tokens")
    contextWindowUsed: int = Field(..., alias="contextWindowUsed", description="Context window utilized")
    promptCacheHits: int = Field(0, alias="promptCacheHits", description="Prompt cache hits")
    costUsd: float = Field(..., alias="costUsd", description="Cost in USD")

    class Config:
        populate_by_name = True


# ============================================================================
# Storage Resource Schemas
# ============================================================================
class StorageResources(BaseModel):
    """Storage resource metrics."""
    tempStorageMb: float = Field(..., alias="tempStorageMb", description="Temporary storage in MB")
    persistentStorageMb: float = Field(..., alias="persistentStorageMb", description="Persistent storage in MB")
    cacheSizeMb: float = Field(..., alias="cacheSizeMb", description="Cache size in MB")
    databaseOperations: int = Field(..., alias="databaseOperations", description="Database operations count")

    class Config:
        populate_by_name = True


# ============================================================================
# Cost Breakdown Schemas
# ============================================================================
class CostBreakdown(BaseModel):
    """Detailed cost breakdown."""
    computeCostUsd: float = Field(..., alias="computeCostUsd", description="Compute costs")
    tokenCostUsd: float = Field(..., alias="tokenCostUsd", description="Token costs")
    apiCostUsd: float = Field(..., alias="apiCostUsd", description="API costs")
    storageCostUsd: float = Field(..., alias="storageCostUsd", description="Storage costs")
    networkCostUsd: float = Field(..., alias="networkCostUsd", description="Network costs")
    totalCostUsd: float = Field(..., alias="totalCostUsd", description="Total costs")

    class Config:
        populate_by_name = True


# ============================================================================
# Resource Utilization Schemas
# ============================================================================
class ResourceUtilizationMetrics(BaseModel):
    """Complete resource utilization metrics."""
    agentId: str = Field(..., alias="agentId")
    executionId: str = Field(..., alias="executionId")
    workspaceId: str = Field(..., alias="workspaceId")
    timestamp: datetime
    computeResources: ComputeResources = Field(..., alias="computeResources")
    llmResources: LLMResources = Field(..., alias="llmResources")
    storageResources: StorageResources = Field(..., alias="storageResources")
    costBreakdown: CostBreakdown = Field(..., alias="costBreakdown")
    executionDurationMs: int = Field(..., alias="executionDurationMs")
    queueWaitTimeMs: int = Field(..., alias="queueWaitTimeMs")
    metadata: Optional[Dict[str, Any]] = None

    class Config:
        populate_by_name = True


# ============================================================================
# API Usage Schemas
# ============================================================================
class APICallStatistics(BaseModel):
    """API call statistics."""
    totalCalls: int = Field(..., alias="totalCalls")
    successfulCalls: int = Field(..., alias="successfulCalls")
    failedCalls: int = Field(..., alias="failedCalls")
    rateLimitedCalls: int = Field(..., alias="rateLimitedCalls")
    timeoutCalls: int = Field(0, alias="timeoutCalls")

    class Config:
        populate_by_name = True


class LatencyMetrics(BaseModel):
    """Latency metrics."""
    avgLatencyMs: float = Field(..., alias="avgLatencyMs")
    p50LatencyMs: float = Field(..., alias="p50LatencyMs")
    p95LatencyMs: float = Field(..., alias="p95LatencyMs")
    p99LatencyMs: float = Field(..., alias="p99LatencyMs")
    minLatencyMs: float = Field(..., alias="minLatencyMs")
    maxLatencyMs: float = Field(..., alias="maxLatencyMs")

    class Config:
        populate_by_name = True


class APICostMetrics(BaseModel):
    """API cost metrics."""
    totalCostUsd: float = Field(..., alias="totalCostUsd")
    costPerCall: float = Field(..., alias="costPerCall")
    wastedCostFailedCalls: float = Field(..., alias="wastedCostFailedCalls")

    class Config:
        populate_by_name = True


class RateLimitInfo(BaseModel):
    """Rate limiting information."""
    currentUsage: int = Field(..., alias="currentUsage")
    limit: int
    resetTime: datetime = Field(..., alias="resetTime")
    throttleIncidents: int = Field(..., alias="throttleIncidents")

    class Config:
        populate_by_name = True


class APIErrorAnalysis(BaseModel):
    """API error analysis."""
    errorTypes: Dict[str, int] = Field(..., alias="errorTypes")
    errorRate: float = Field(..., alias="errorRate")
    retrySuccessRate: float = Field(..., alias="retrySuccessRate")

    class Config:
        populate_by_name = True


class APIUsageMetrics(BaseModel):
    """Comprehensive API usage metrics."""
    agentId: str = Field(..., alias="agentId")
    apiEndpoint: str = Field(..., alias="apiEndpoint")
    apiProvider: Optional[str] = Field(None, alias="apiProvider")
    timePeriod: str = Field(..., alias="timePeriod")
    usageStats: APICallStatistics = Field(..., alias="usageStats")
    latencyMetrics: LatencyMetrics = Field(..., alias="latencyMetrics")
    costMetrics: APICostMetrics = Field(..., alias="costMetrics")
    rateLimiting: RateLimitInfo = Field(..., alias="rateLimiting")
    errorAnalysis: APIErrorAnalysis = Field(..., alias="errorAnalysis")

    class Config:
        populate_by_name = True


# ============================================================================
# Token Analytics Schemas
# ============================================================================
class TokenDistribution(BaseModel):
    """Token distribution analysis."""
    inputTokenPercent: float = Field(..., alias="inputTokenPercent")
    outputTokenPercent: float = Field(..., alias="outputTokenPercent")
    embeddingTokenPercent: float = Field(..., alias="embeddingTokenPercent")

    class Config:
        populate_by_name = True


class TokenEfficiencyMetrics(BaseModel):
    """Token efficiency metrics."""
    tokensPerExecution: float = Field(..., alias="tokensPerExecution")
    tokensPerDollar: float = Field(..., alias="tokensPerDollar")
    cacheHitRate: float = Field(..., alias="cacheHitRate")
    contextWindowUtilization: float = Field(..., alias="contextWindowUtilization")

    class Config:
        populate_by_name = True


class PromptOptimization(BaseModel):
    """Prompt optimization suggestions."""
    redundantTokens: int = Field(..., alias="redundantTokens")
    compressionRatio: float = Field(..., alias="compressionRatio")
    savingsPercent: float = Field(..., alias="savingsPercent")
    costSavingsUsd: float = Field(..., alias="costSavingsUsd")

    class Config:
        populate_by_name = True


class TokenAnalysis(BaseModel):
    """Comprehensive token analysis."""
    totalTokens: int = Field(..., alias="totalTokens")
    totalCostUsd: float = Field(..., alias="totalCostUsd")
    distribution: TokenDistribution
    efficiency: TokenEfficiencyMetrics
    optimization: PromptOptimization

    class Config:
        populate_by_name = True


# ============================================================================
# Budget Schemas
# ============================================================================
class TokenBudget(BaseModel):
    """Token budget configuration."""
    workspaceId: str = Field(..., alias="workspaceId")
    agentId: Optional[str] = Field(None, alias="agentId")
    dailyTokenBudget: Optional[int] = Field(None, alias="dailyTokenBudget")
    weeklyTokenBudget: Optional[int] = Field(None, alias="weeklyTokenBudget")
    monthlyTokenBudget: Optional[int] = Field(None, alias="monthlyTokenBudget")
    dailyCostBudgetUsd: Optional[float] = Field(None, alias="dailyCostBudgetUsd")
    weeklyCostBudgetUsd: Optional[float] = Field(None, alias="weeklyCostBudgetUsd")
    monthlyCostBudgetUsd: Optional[float] = Field(None, alias="monthlyCostBudgetUsd")
    warningThresholdPercent: int = Field(80, alias="warningThresholdPercent")
    criticalThresholdPercent: int = Field(90, alias="criticalThresholdPercent")

    class Config:
        populate_by_name = True


class BudgetUsage(BaseModel):
    """Budget usage tracking."""
    dailyTokens: int = Field(..., alias="dailyTokens")
    dailyCostUsd: float = Field(..., alias="dailyCostUsd")
    dailyBudgetUsagePercent: float = Field(..., alias="dailyBudgetUsagePercent")
    dailyCostBudgetUsagePercent: float = Field(..., alias="dailyCostBudgetUsagePercent")
    rolling7dTokens: int = Field(..., alias="rolling7dTokens")
    rolling30dTokens: int = Field(..., alias="rolling30dTokens")
    rolling30dCostUsd: float = Field(..., alias="rolling30dCostUsd")
    budgetStatus: BudgetStatus = Field(..., alias="budgetStatus")

    class Config:
        populate_by_name = True


# ============================================================================
# Optimization Schemas
# ============================================================================
class OptimizationRecommendation(BaseModel):
    """Resource optimization recommendation."""
    id: str
    agentId: str = Field(..., alias="agentId")
    workspaceId: str = Field(..., alias="workspaceId")
    optimizationType: str = Field(..., alias="optimizationType")
    category: OptimizationCategory
    priority: Priority
    title: str
    description: str
    reasoning: str
    currentValue: Optional[float] = Field(None, alias="currentValue")
    recommendedValue: Optional[float] = Field(None, alias="recommendedValue")
    estimatedSavingsUsd: float = Field(..., alias="estimatedSavingsUsd")
    estimatedSavingsPercent: float = Field(..., alias="estimatedSavingsPercent")
    implementationEffort: ImplementationEffort = Field(..., alias="implementationEffort")
    implementationSteps: Optional[List[str]] = Field(None, alias="implementationSteps")
    status: str
    createdAt: datetime = Field(..., alias="createdAt")

    class Config:
        populate_by_name = True


# ============================================================================
# Waste Detection Schemas
# ============================================================================
class WasteEvent(BaseModel):
    """Resource waste event."""
    id: str
    agentId: str = Field(..., alias="agentId")
    executionId: Optional[str] = Field(None, alias="executionId")
    workspaceId: str = Field(..., alias="workspaceId")
    wasteType: WasteType = Field(..., alias="wasteType")
    wasteCategory: str = Field(..., alias="wasteCategory")
    title: str
    description: str
    wasteAmount: float = Field(..., alias="wasteAmount")
    wasteUnit: str = Field(..., alias="wasteUnit")
    wasteCostUsd: float = Field(..., alias="wasteCostUsd")
    confidenceScore: float = Field(..., alias="confidenceScore")
    isResolved: bool = Field(False, alias="isResolved")
    detectedAt: datetime = Field(..., alias="detectedAt")

    class Config:
        populate_by_name = True


class WasteSummary(BaseModel):
    """Waste detection summary."""
    totalWasteCostUsd: float = Field(..., alias="totalWasteCostUsd")
    wasteByType: Dict[str, float] = Field(..., alias="wasteByType")
    unresolvedCount: int = Field(..., alias="unresolvedCount")
    potentialMonthlySavings: float = Field(..., alias="potentialMonthlySavings")

    class Config:
        populate_by_name = True


# ============================================================================
# Forecast Schemas
# ============================================================================
class ResourceForecast(BaseModel):
    """Resource demand forecast."""
    resourceType: ResourceType = Field(..., alias="resourceType")
    forecastHorizonDays: int = Field(..., alias="forecastHorizonDays")
    predictedValue: float = Field(..., alias="predictedValue")
    lowerBound: float = Field(..., alias="lowerBound")
    upperBound: float = Field(..., alias="upperBound")
    confidenceLevel: float = Field(..., alias="confidenceLevel")
    historicalAverage: float = Field(..., alias="historicalAverage")
    historicalTrend: str = Field(..., alias="historicalTrend")
    exceedsBudget: bool = Field(..., alias="exceedsBudget")
    forecastDate: date = Field(..., alias="forecastDate")

    class Config:
        populate_by_name = True


class CostProjection(BaseModel):
    """Cost projection analysis."""
    projectedDailyCost: float = Field(..., alias="projectedDailyCost")
    projectedWeeklyCost: float = Field(..., alias="projectedWeeklyCost")
    projectedMonthlyCost: float = Field(..., alias="projectedMonthlyCost")
    growthRate: float = Field(..., alias="growthRate")
    budgetAlerts: List[str] = Field(..., alias="budgetAlerts")

    class Config:
        populate_by_name = True


# ============================================================================
# Efficiency Schemas
# ============================================================================
class EfficiencyScore(BaseModel):
    """Agent efficiency scoring."""
    overallScore: float = Field(..., alias="overallScore", description="Overall efficiency score 0-100")
    tokensPerDollar: float = Field(..., alias="tokensPerDollar")
    executionsPerDollar: float = Field(..., alias="executionsPerDollar")
    throughputScore: float = Field(..., alias="throughputScore")
    costEfficiencyPercent: float = Field(..., alias="costEfficiencyPercent")
    percentileRank: Optional[int] = Field(None, alias="percentileRank")
    improvementAreas: List[str] = Field(..., alias="improvementAreas")

    class Config:
        populate_by_name = True


# ============================================================================
# Response Schemas
# ============================================================================
class ResourceAnalyticsResponse(BaseModel):
    """Complete resource analytics response."""
    agentId: str = Field(..., alias="agentId")
    workspaceId: str = Field(..., alias="workspaceId")
    timePeriod: str = Field(..., alias="timePeriod")
    resourceMetrics: ResourceUtilizationMetrics = Field(..., alias="resourceMetrics")
    tokenAnalysis: TokenAnalysis = Field(..., alias="tokenAnalysis")
    costBreakdown: CostBreakdown = Field(..., alias="costBreakdown")
    efficiencyScore: EfficiencyScore = Field(..., alias="efficiencyScore")
    budgetUsage: Optional[BudgetUsage] = Field(None, alias="budgetUsage")
    optimizations: List[OptimizationRecommendation] = []
    wasteDetection: WasteSummary = Field(..., alias="wasteDetection")
    forecasts: List[ResourceForecast] = []

    class Config:
        populate_by_name = True


class CostAnalysisResponse(BaseModel):
    """Cost analysis response."""
    workspaceId: str = Field(..., alias="workspaceId")
    period: str
    totalCost: float = Field(..., alias="totalCost")
    costByAgent: Dict[str, float] = Field(..., alias="costByAgent")
    costBreakdown: CostBreakdown = Field(..., alias="costBreakdown")
    costTrends: List[Dict[str, Any]] = Field(..., alias="costTrends")
    projections: CostProjection
    optimizationPotential: float = Field(..., alias="optimizationPotential")

    class Config:
        populate_by_name = True
